Add new filtered deck options implementation

lukstbit opened the pull request:
## Purpose / Description

Initially, while working on #18875, I implemented the filtered deck related part outside of the (old)_FilteredDeckOptions_ like _CreateDeckDialog_ was doing. After thinking about this I decided that it is a better idea to implement a new screen for filtered decks following desktop behavior.

Changes:
- use material design for screen
- get away from deprecated preferences code
- automatic configuration changes handling
- slightly improve user ux as entering most values requires only two touches now
- still a bit more functionality to add to match desktop(browser search). I implemented it but this already got really big

The line count is big but two thirds of it is the layout, tests and views setup.

Video with general behavior(Don't keep activities is on):

[Screen_recording_20251202_092039.webm](https://github.com/user-attachments/assets/1d0bd71b-131c-4790-b37d-7856cb40b670)



<details>
<summary>Images with how it looks with different themes</summary>

<img width="540" height="1200" alt="Screenshot_20251206_165459" src="https://github.com/user-attachments/assets/f72ac34f-51dc-4aa6-bd62-adc849fb5800" />
<img width="540" height="1200" alt="Screenshot_20251206_165527" src="https://github.com/user-attachments/assets/d99d3b1e-6d55-4da0-b3c7-adc96d73baea" />
<img width="540" height="1200" alt="Screenshot_20251206_165545" src="https://github.com/user-attachments/assets/f6fa9be9-6042-4b0c-b5a0-9e7f09057af3" />



</details>

## Fixes
* Related to #18875

## How Has This Been Tested?

Manually checked the behavior against desktop behavior, ran tests.

## Checklist

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

lukstbit committed:
Add url constant for ankiweb filtered decks related docs

To be used in filtered decks options screen.

lukstbit committed:
Add ConfigureAwareSingleFragmentActivity

Uses the implementation of SingleFragmentActivity and delegates the
configuration changes back to the system.

Handling configuration changes on our own is not something useful for most
fragments(excluding page fragment which are mostly a WebView) and has the
potential to introduce quite a few bugs.

lukstbit committed:
Add implementation for new filtered deck options screen

Note: this doesn't implement completely the desktop functionality for
this screen, everything missing will be added in a future PR.

lukstbit committed:
Use new filtered deck options code

lukstbit committed:
Remove previous filtered deck options related code

ZornHadNoChoice commented:
If I may give my feedback:
1. The titles "Filter" and "Options" should be a bit more distinguished from the other text.
2. The sort selection box can be typed in?
3. [Use switches (not radio buttons) if the items in a list can be independently controlled](https://m3.material.io/#:~:text=Use%20switches%20(not%20radio%20buttons)%20if%20the%20items%20in%20a%20list%20can%20be%20independently%20controlled).

ZornHadNoChoice commented:
I think the remaining checkboxes should also be switches. Also, it would be better if all switches were below "Options", the switches font were normal weight, and the second filter/delays were above it.

<details><summary>If you're fine with deviating slightly from Anki Desktop</summary>
<p>
the "Name" box is unnecessary and the "Build" button could be replaced with a checkmark icon.
</p>
</details>

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserFragment.kt:
```diff
@@ -771,24 +772,44 @@ class CardBrowserFragment :
+     */
+    private suspend fun buildDeckNameSearch(): String? =
+        if (activityViewModel.deckId == ALL_DECKS_ID) {
+            "deck:_*"
```
What's the `_` for here?

lukstbit replied:
This is the query used on desktop if you go to browser and select "Decks" from the side menu.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsState.kt:
```diff
@@ -0,0 +1,110 @@
+@Parcelize
+data class FilteredDeckOptions(
+    /** If 0 we are creating a deck otherwise it's the [DeckId] of a filtered deck */
+    val id: DeckId = 0,
```
can `null` be used here?

david-allison commented on AnkiDroid/src/test/java/com/ichi2/anki/filtered/FilteredDeckOptionsViewModelTest.kt:
```diff
@@ -0,0 +1,207 @@
+            assertThat(deck.name, equalTo("TestFiltered"))
+            val count = withCol { decks.cardCount(deck.id) }
+            assertThat(count, equalTo(2))
+            assertEquals(deck.id, n1.cards()[0].did)
```
I think `firstCard()` is usable

david-allison reviewed (changes requested):
Wow, excellent!

lukstbit commented on AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserFragment.kt:
```diff
@@ -771,24 +772,44 @@ class CardBrowserFragment :
+        } else {
+            activityViewModel.deckId?.let {
+                // decks.name() returns 'no deck' if a deck doesn't exist for that did
+                "\"deck:${withCol { decks.get(it, default = true)?.name }}\""
```
The entire _deck: deck_name_ construct is wrapped in quotes just like desktop does.

lukstbit commented on AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsFragment.kt:
```diff
@@ -0,0 +1,306 @@
+                    cardOptions,
+                ),
+            )
+            setText(cardOptions[selectedIndex], false)
```
Can't, not a kotlin method.

lukstbit commented on AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsState.kt:
```diff
@@ -0,0 +1,110 @@
+ * State that indicates the successful building/rebuilding of a filtered deck. After receiving this
+ * the view is expected to exit the screen.
+ */
+data object Finished : FilteredDeckOptionsState
```
Went with _DeckBuilt_, I liked the proposal.

lukstbit commented on AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsViewModel.kt:
```diff
@@ -0,0 +1,326 @@
+        limit: String,
+    ) {
+        Timber.i("Filtered deck filter($index) limit changing to $limit")
+        // TODO maybe export input errors so we can display them with the field
```
Implemented input feedback for when the input boxes for limits and reschedule delays are empty/invalid.

lukstbit commented on AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsViewModel.kt:
```diff
@@ -0,0 +1,326 @@
+
+    /** Saves to the [SavedStateHandle] the result of invoking [action] and after updates the state. */
+    private fun updateCurrentState(action: FilteredDeckOptions.() -> FilteredDeckOptions) {
+        val updatedState = currentState().action()
```
?
_currentState()_ it's just a getter to make working with the SavedStateHandle easier I don't think this is important.

lukstbit commented:
I updated the images.

> I think the remaining checkboxes should also be switches.

Tried it and like it enough to implement it, thanks.

> Also, it would be better if all switches were below "Options", the second filter/delays were above it, and the switches font were normal weight.

Did this on purpose. 
The second filter switch I placed outside because I don't think the desktop behavior is that great here(switch in a different place => content appearing somewhere above it). I like the local aspect of the current behavior.
The delays inputs are only visible when reschedule is unchecked and again I like the local aspect of it.

> the "Name" box is unnecessary 

Probably but it's also consistent with all the other inputs so I'm keeping it.

> and the "Build" button could be replaced with a checkmark icon. 

Doesn't work, we have two cases Build and Rebuild and the text is important.

> Turn off spellchecking in the dropdown
> Can you use app:suffixText (possibly with a <plurals> here):

Implemented both requests, but note that I used the backend string for seconds.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsFragment.kt:
```diff
@@ -0,0 +1,332 @@
+
+    private fun bindListeners() {
+        binding.deckNameInput.doOnTextChanged { text, _, _, _ ->
+            if (text == null) return@doOnTextChanged
```
This is such a common pattern, it might be worth a local extension method

david-allison reviewed (approved):
Looks great

lukstbit commented:
Note that I marked the test: _building a filtered deck with no cards fails unless allow empty is checked_ as flaky on windows. Passed on linux every time I ran the test but in Windows CI it sometimes fails.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsFragment.kt:
```diff
@@ -0,0 +1,329 @@
+        binding.filterLimitInputLayout.error =
+            if (filter1State.limit.toIntOrNull() == null) TR.errorsInvalidInputEmpty() else null
+        binding.filterCardsInput.setAdapterIfChanged(state.cardOptions, filter1State.index)
+        // rescheduling(done here because in filter 2 setup we might exit early)
```
```suggestion
        // rescheduling (done here because in filter 2 setup we might exit early)
```

david-allison reviewed (approved):
All optional/nitpicks. Looks GREAT!!!

BrayanDSO commented:
the first input is not aligned with the others

<img width="385" height="517" alt="image" src="https://github.com/user-attachments/assets/cbd35d5f-7206-4a63-8539-03b78d8b5203" />

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsFragment.kt:
```diff
@@ -0,0 +1,329 @@
+import kotlinx.coroutines.launch
+
+/**
+ * Represents the screen where a filtered deck can be built or rebuilt after updating it's properties.
```
```suggestion
 * Represents the screen where a filtered deck can be built or rebuilt after updating its properties.
```

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsFragment.kt:
```diff
@@ -0,0 +1,329 @@
+
+/**
+ * Represents the screen where a filtered deck can be built or rebuilt after updating it's properties.
+ *
```
Could you please document the behavior of the two loading indicators?

I had to check in the code what each of them do.

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsViewModel.kt:
```diff
@@ -0,0 +1,324 @@
+    }
+
+    /** Queries the backend for cards search options and [FilteredDeckForUpdate] guarding against any exception. */
+    private fun Collection.safeBackendDataQuery(did: DeckId): Result<Pair<FilteredDeckForUpdate, List<String>>> =
```
Document filtered decks backend data query exceptions

`safeBackendDataQuery()` and `safeAddOrUpdateFilteredDeck()` guard against errors while building a filtered deck.

The errors that can be thrown should be documented so other developers know what they should expect

---
text to peel to a new issue ⤴️

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsViewModel.kt:
```diff
@@ -0,0 +1,324 @@
+                            Deck.Filtered.SearchTerm.Order
+                                .forNumber(filter2State.index),
+                        ),
+                )
```
searchTerms got better, IMO

Deck.Filtered may be cleaner as a builder instead of using dsl

```patch
Subject: [PATCH] a
---
Index: AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsViewModel.kt b/AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsViewModel.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsViewModel.kt	(revision 6b787c2253308cb9858ed7137b9dcf103f6822fb)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/filtered/FilteredDeckOptionsViewModel.kt	(date 1768596709543)
@@ -21,7 +21,10 @@
 import androidx.lifecycle.viewModelScope
 import anki.collection.OpChangesWithId
 import anki.decks.Deck
+import anki.decks.DeckKt.FilteredKt.searchTerm
+import anki.decks.DeckKt.filtered
 import anki.decks.FilteredDeckForUpdate
+import anki.decks.copy
 import anki.decks.filteredDeckForUpdate
 import com.ichi2.anki.CollectionManager.withCol
 import com.ichi2.anki.filtered.FilteredDeckOptionsFragment.Companion.ARG_DECK_ID
@@ -284,37 +287,32 @@
             id = current.id ?: 0
             name = current.name
             allowEmpty = current.allowEmpty
-            val configBuilder =
-                Deck.Filtered
-                    .newBuilder()
-                    .setReschedule(current.shouldReschedule)
-                    .setPreviewAgainSecs(current.delayAgain.toIntOrNull() ?: 60)
-                    .setPreviewHardSecs(current.delayHard.toIntOrNull() ?: 600)
-                    .setPreviewGoodSecs(current.delayGood.toIntOrNull() ?: 0)
-                    .addSearchTerms(
-                        Deck.Filtered.SearchTerm
-                            .newBuilder()
-                            .setSearch(current.filter1State.search)
-                            .setLimit(current.filter1State.limit.toIntOrNull() ?: 0)
-                            .setOrder(
-                                Deck.Filtered.SearchTerm.Order
-                                    .forNumber(current.filter1State.index),
-                            ),
+            val config =
+                filtered {
+                    reschedule = current.shouldReschedule
+                    previewAgainSecs = current.delayAgain.toIntOrNull() ?: 60
+                    previewHardSecs = current.delayHard.toIntOrNull() ?: 600
+                    previewGoodSecs = current.delayGood.toIntOrNull() ?: 0
+                    searchTerms.add(
+                        searchTerm {
+                            search = current.filter1State.search
+                            limit = current.filter1State.limit.toIntOrNull() ?: 0
+                            order = Deck.Filtered.SearchTerm.Order.forNumber(current.filter1State.index)
+                        }
                     )
+                }
             val filter2State = current.filter2State
             if (current.isSecondFilterEnabled && filter2State != null) {
-                configBuilder.addSearchTerms(
-                    Deck.Filtered.SearchTerm
-                        .newBuilder()
-                        .setSearch(filter2State.search)
-                        .setLimit(filter2State.limit.toIntOrNull() ?: 0)
-                        .setOrder(
-                            Deck.Filtered.SearchTerm.Order
-                                .forNumber(filter2State.index),
-                        ),
-                )
+                config.copy {
+                    searchTerms.add(searchTerm {
+                        search = filter2State.search
+                        limit = filter2State.limit.toIntOrNull() ?: 0
+                        order = Deck.Filtered.SearchTerm.Order
+                            .forNumber(filter2State.index)
+                    })
+                }
             }
-            config = configBuilder.build()
+            this.config = config
         }
 
     companion object {
```

BrayanDSO reviewed (changes requested):
Currently, the only blocking thing for me is the input boxes misalignment
